### PR TITLE
Remove init func in install.go

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix commit is https://github.com/kubernetes/kubernetes/pull/42722

This allows us to move the apis to a new repo without copying global variables like api.Scheme, api.Registry, api.GroupFactoryRegistry.

ref https://github.com/kubernetes/kubernetes/issues/44065.

### CODE REVIEW

**Comment by MHBauer** [comment]
This would be amazing. I definitely favor "explicitly install" over the "spooky action at a distance anonymous import".

**Comment by caesarxuchao** [comment]
The e2e test failure was caused by kube-proxy, which was relying on the side-effect of importing "k8s.io/kubernetes/apis/componentconfig/install". It's fixed now.

**Comment by caesarxuchao** [comment]
@lavalamp could you take a look at the second commit? Thanks.

File: pkg/apis/settings/install/install.go [hunk 1]
```diff
@@ -22,17 +22,15 @@ import (
...
 	"k8s.io/kubernetes/pkg/apis/settings/v1alpha1"
 )
 
-func init() {
-	Install(api.GroupFactoryRegistry, api.Registry, api.Scheme)
-}
-
 // Install registers the API group and adds types to a scheme
 func Install(groupFactoryRegistry announced.APIGroupFactoryRegistry, registry *registered.APIRegistrationManager, scheme *runtime.Scheme) {
+	if _, ok := groupFactoryRegistry[settings.GroupName]; ok {
```
Comment (sttts) [inline pkg/apis/settings/install/install.go]: Can't we move these checks into `Announce()` and `RegisterAndEnable` by making them idem-potent? These three lines are ugly.

Comment (caesarxuchao) [inline pkg/apis/settings/install/install.go]: @sttts i addressed the comment in the last commit. PTAL, thanks.

**Comment by MHBauer** [comment]
For what it's worth, I like the way this looks.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go [hunk 2]
```diff
@@ -89,6 +89,17 @@ func (gmf *GroupMetaFactory) Announce(groupFactoryRegistry APIGroupFactoryRegist
...
 	}
 	return gmf
+
+}
+
+// AnnounceIfNotAnnounced is the same as Announce, except that it returns nil
+// if gmf.GroupArgs.GroupName is already registered in groupFactoryRegistry,
+// instead of panicing.
+func (gmf *GroupMetaFactory) AnnounceIfNotAnnounced(groupFactoryRegistry APIGroupFactoryRegistry) *GroupMetaFactory {
+	if _, ok := groupFactoryRegistry[gmf.GroupArgs.GroupName]; ok {
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: one could put some concistency checks here, i.e. if you re-announce with different values. Just an idea...

File: federation/cmd/federation-apiserver/app/import_known_versions.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	"k8s.io/kubernetes/federation/apis/core"
+	coreinstall "k8s.io/kubernetes/federation/apis/core/install"
+	federationinstall "k8s.io/kubernetes/federation/apis/federation/install"
+	"k8s.io/kubernetes/pkg/api"
+	autoscalinginstall "k8s.io/kubernetes/pkg/apis/autoscaling/install"
+	batchinstall "k8s.io/kubernetes/pkg/apis/batch/install"
+	extensionsinstall "k8s.io/kubernetes/pkg/apis/extensions/install"
+)
+
+func init() {
```
Comment (sttts) [inline federation/cmd/federation-apiserver/app/import_known_versions.go]: Can we move these install calls where the scheme+registery are actually declared?

Comment (sttts) [inline federation/cmd/federation-apiserver/app/import_known_versions.go]: On that way, we might hit import cycles which are worth cutting. At least, we get away one more step from global schemes by that.

Comment (caesarxuchao) [inline federation/cmd/federation-apiserver/app/import_known_versions.go]: Federation is using the global api.GroupFactoryRegistry/Registry/Scheme. I agree they should declare their own ones and these calls can then be moved to where the scheme/registry are declared. However, I tend to leave that to another PR. In this PR, i just want to cut the dependency of `install` packages on these global variables.

Comment (lavalamp) [inline federation/cmd/federation-apiserver/app/import_known_versions.go]: Yeah, this is not the final state, we're trying to do this in smallish easy to verify chunks.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go [hunk 4]
```diff
@@ -89,6 +89,17 @@ func (gmf *GroupMetaFactory) Announce(groupFactoryRegistry APIGroupFactoryRegist
 		panic(err)
 	}
 	return gmf
+
+}
+
+// AnnounceIfNotAnnounced is the same as Announce, except that it returns nil
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: This is the only thing I don't like in this PR, why does this change cause double-registrations now?

Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: For a binary like the apiserver, it imports both pkg/master and pkg/client, which have their own import_known_versions.go that call Install(), thus the double-registrations.

The original code doesn't suffer from double-registrations because the `Install()` is called in the init function of each API groupVersion. Though the `install` packages might be imported multiple times, the init() function is guaranteed to be run only once.

Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: I see. I'm trying to think of a way to avoid this. Maybe we can keep a sync.Once variable near where the three things are declared, like `var APITypeInstall sync.Once` and then client/server/test code can use that to wrap their install?

Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: Hmm the goal is "installing a specific API type in an instance of scheme/registry/APIGroupFactoryRegistry only once", i don't think we can define a sync.Once to achieve this goal.

I think @sttts 's suggestion https://github.com/kubernetes/kubernetes/pull/44171#discussion_r110819851 might be better.

Comment (MHBauer) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: If Announce suffers from a race/double import, it's probably better to fix it in Announce rather than add a new way to do things. If not, Announce needs a warning about not calling it twice.

Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: it's more that I want users to take care announcing/registering api groups and not randomly call them in a bunch of places "to make sure". I'd rather have a way to import the client that allows you to manually register. I know this is an intermediate state and once we give the client its own copy of the groups, we won't have this problem any more, but I worry that if we stay in the intermediate state for more than a few hours people will add double registrations all over and it will be hard to clean up the mess.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 2 above)
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: it should be very predictable what groups are going to get registered by any one binary. Tolerating & recovering if people register multiple times will cause the code to be less understandable in the long run.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 4 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: I'll try to eliminate the double registration if that's not too messy. Client have actually defined its own copy of scheme/registry/APIGroupFactoryRegistry in client-go/kubernetes/scheme, but i guess they are used in mix with api.Scheme.

Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: Ok, it seems i cannot eliminate the double-registrations yet. Because client-go/pkg/api/v1/helpers.go and reousrce_helper.go still depend on api.Scheme, and because that code is copied instead of being authoratative, i cannot rewrite it to client-go/kubernetes/scheme.Scheme.

With my other refactoring, client-go won't have those two files. After those refactoring are merged, i'll try again to eliminate the double registration.

File: federation/cmd/federation-apiserver/app/import_known_versions.go (hunk 3 above)
Comment (sttts) [inline federation/cmd/federation-apiserver/app/import_known_versions.go]: I agree, follow-up is fine.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 2 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: I totally agree after we move the install calls next to Scheme+Registry, i.e. same file. We had a similar situation with registered types, i.e. registering same type with multiple GVs. In theory it's predictable, in practice we had two handful of inconcistencies.

File: staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go (hunk 4 above)
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go]: Eliminating double registration is a much bigger refactor than i expected. I had refactored client-go to not relying on api.Scheme/Registry/Codecs (https://github.com/kubernetes/kubernetes/pull/44523), but other packages like kubectl, testapi, federation and more are still relying on api.XXX, and they rely on multiple registration to make sure the API are registered.

File: pkg/apis/settings/install/install.go (hunk 1 above)
Comment (caesarxuchao) [inline pkg/apis/settings/install/install.go]: @lavalamp @sttts i reverted to this commit.

At this version, install.go returns early if the group is already announced. I think this is slightly better than adding an `AnnounceIfNotAnnounced` function, because it keeps the `announced` package clean.

This PR is not making the multi-registration problem worse, people are already doing multiple-registration and we were relying on `init()` to hide that fact. Completely removing multi-registration is a too big refactor than i can handle right now.

Please let me know if you think the PR is good enough.

**Comment by MHBauer** [comment]
Does this have a chance to go in? I still think we should not be using `init` like this.

**Comment by caesarxuchao** [comment]
Maybe after https://github.com/kubernetes/kubernetes/pull/44784. Don't want to do two big refactors at the same time.
